pylint/checkers/format.py

```diff
@@ -676,7 +667,14 @@ def check_lines(self, lines: str, lineno: int) -> None:
         split_lines = self.specific_splitlines(lines)
 
         for offset, line in enumerate(split_lines):
-            self.check_line_ending(line, lineno + offset)
+            if not line.endswith("\n"):
```

Comment:
It seems like the check is done on every line, but it could be done only on the last line ? Also what about cross platform line ending (``\r\n```and ```\r``) for Windows and Mac ?

Reply:
I'm not sure. I just had to extract this out of the method because there was no simple way to not call the check of `trailing-whitespace`. I'll try out your suggestion but since we don't have any issues about this I don't think we should also change the line endings that are checked in this PR.

Reply:
Oh never mind.

I can't extract this. We need to be able to skip the `trailing-whitespace` check if we added the `missing-final-newline`. We can do something with a `emitted_final_newline` variable but I think checking `if emitted_final_newline` is as quick as `if line.endswith("\n")` and the current code is a bit clearer.

Reply:
> We need to be able to skip the trailing-whitespace check if we added the missing-final-newline.

How about we loop on ``split_lines[:-1]`` and handle the last line separately ? I don't like knowingly making a check O(n) when it could be O(1)

Reply:
I don't think there is a good way without duplicating a big part of code + the comment? There is also the option to do something like `offset == len(split_lines)` but then again, I'm not sure an `endswith` is that much slower.

Reply:
What about something like this ?
```python
for offset, line in enumerate(split_lines[-1]):
    self.check_trailing_whitespace_ending(last_line, lineno + offset, line_start):
last_line =  split_lines[-1][1]
offset =  split_lines[-1][0]
if not last_line.endswith("\n"):
    self.add_message("missing-final-newline", line=lineno + offset)
else:
    self.check_trailing_whitespace_ending(last_line, lineno + offset, line_start):

def check_trailing_whitespace_ending(...)
    # We don't test for trailing whitespaces in strings
    # See https://github.com/PyCQA/pylint/issues/6936
    # and https://github.com/PyCQA/pylint/issues/3822
    if tokens.type(line_start) == tokenize.STRING:
        return
    ...
```

Reply:
This also doesn't work since `split_lines` isn't always full, so `split_lines[-1]` can return an `IndexError` 😠 

I don't think there is anyway to do this without adding additional if statements and assignments which kind of defeat the purpose of calling `endswith` once.